numpy/core/src/multiarray/einsum.c.src

```diff
@@ -2740,7 +2740,7 @@ PyArray_EinsteinSum(char *subscripts, npy_intp nop,
         op_axes[nop][idim] = idim;
     }
     for (idim = ndim_output; idim < ndim_iter; ++idim) {
-        op_axes[nop][idim] = -1;
+        op_axes[nop][idim] = NPY_ITER_REDUCTION_AXIS(-1);
```

Comment:
Will this change make `np.einsum("...i,...i->...i", np.arange(3), np.array([1]))` an error? That is, should named dimensions in einsum be viewed as non-broadcastable?

Reply:
Yes, it changes that. I can undo. But einsum does not really have `keepdims` semantics nor is that notation einsum notation AFAIK.
Completely changing the resul behaviour by providing (a potentially wrong) output array seems undesirable to me?

Reply:
What's the resulting error look like with this branch?

Not sure what you mean by that but behind einsum notation - that notation works for me

Reply:
Sorry, this is unchanged, I thought you had `np.einsum("...i->...i", np.arange(3), out=np.array([1]))`

If you do not provide an out array, no semantics are ever changed here.

Reply:
OK, sorry... I do not think anything changed at all. Since opposite broadcasting should be the only thing that changed (since only `np.newaxis` is marked here). But einsum specifically checks for `out` having too many dimensions.

Reply:
I.e. this is really just because I like this way better to specify the axes. The only thing where it makes a real difference currently would be reductions with `keepdims=True`, but there I think we may have had an additional check recently. Making the only actual bug fix/change in behaviour gufuncs...

Reply:
So what threw me off a bit, is that this does *not* address the fact that the output can cause normal expand the final result shape through broadcasting:
```
np.sum(arr, out=np.empty((3, 1)), axis=1, keepdims=True)
```
the output (writeable) array can modify the broadcast dimensions. That is actually a small regression here right now.

Reply:
For reductions, I can fix this trivially by using the no-broadcast flags on the input, which is obviously correct. slightly less nice error:
```
ValueError: non-broadcastable operand with shape (1,100) doesn't match the broadcast shape (3,100)
```
but it seems OK for me for now. For einsum its more complicated, since what we really want is the output to not have any say in the final shape, but there is no nice way to do that currently. However, there is also no change in behaviour here, so I think it does not matter for this PR.

Reply:
That sounds good. But as a check, would one or the other solve the no-input-but-have-output case discussed in https://github.com/numpy/numpy/issues/15140?

Reply:
OK, I had a quick look and this PR does not address #15140, so no worries about it!

Reply:
Hmmmm. I am wondering if we could mark an operand as output to indicate that it should not affect shape (or maybe using the WRITABLE flag is OK, but I fear that is a bit of a potential regression).

I suppose when there is no shape, the shape is just should be `()` (and if there are reduce, but keepdims, axis, they are always inserted with size 1 right now).

With your gufunc the question is whether you actually *want* the output to define the size in that case. Technically, that could be nice in some cases, unless you argue that the user should somehow provide `shape=...` in that case.
Although, for the iterator maybe that would be the same, and the gufunc machinery would just not pass the flag if there are no inputs?

Reply:
Oh, I am conflating two issues there. With no arguments we could already simply default to a 0-D iterator without much hassle probably...
How/if the output argument can affect the iterator shape is another issue.

Reply:
Indeed, for my strange case I do not think the output shape has to affect the iterator; it was about automatically allocating gufunc dimensions if needed, even if there are no inputs.